src/main/java/gregtech/common/covers/CoverFluidRegulator.java

```diff
@@ -86,48 +97,53 @@ protected int doTransferExact(int transferLimit, IFluidHandler sourceHandler, IF
 
     /**
      * Performs one tick worth of Keep Exact behavior.
+     *
      * @param transferLimit the maximum amount in milliBuckets that may be transferred in one tick
      * @param sourceHandler source(s) to move fluids from
-     * @param destHandler destination(s) to move fluids to
-     * @param fluidFilter a predicate which determines what fluids may be moved
-     * @param keepAmount the desired amount in milliBuckets of a particular fluid in the destination
+     * @param destHandler   destination(s) to move fluids to
+     * @param fluidFilter   a predicate which determines what fluids may be moved
      * @return the total amount in milliBuckets of all fluids transferred from source to dest by this method
      */
     protected int doKeepExact(final int transferLimit,
                               final IFluidHandler sourceHandler,
                               final IFluidHandler destHandler,
                               final Predicate<FluidStack> fluidFilter,
-                              final int keepAmount) {
-        if(sourceHandler == null || destHandler == null || fluidFilter == null || keepAmount <= 0)
+                              int keepAmount) {
+
+        if (sourceHandler == null || destHandler == null || fluidFilter == null)
             return 0;
 
         final Map<FluidStack, Integer> sourceFluids =
-            collectDistinctFluids(sourceHandler, IFluidTankProperties::canDrain, fluidFilter);
+                collectDistinctFluids(sourceHandler, IFluidTankProperties::canDrain, fluidFilter);
         final Map<FluidStack, Integer> destFluids =
-            collectDistinctFluids(destHandler, IFluidTankProperties::canFill, fluidFilter);
+                collectDistinctFluids(destHandler, IFluidTankProperties::canFill, fluidFilter);
 
         int transferred = 0;
-        for(FluidStack fluidStack : sourceFluids.keySet()) {
-            if(transferred >= transferLimit)
+        for (FluidStack fluidStack : sourceFluids.keySet()) {
+            if (transferred >= transferLimit)
                 break;
 
+            if (this.fluidFilter.getFilterWrapper().getFluidFilter() != null) {
+                keepAmount = this.fluidFilter.getFilterWrapper().getFluidFilter().getFluidTransferLimit(fluidStack);
```

Comment:
Should this be the maximum/minimum between the transfer limit and the passed keep amount?

Reply:
keepAmount should be the global if the mode is not keep exact/supply or overflown